File: server/internal/crypto/vectors_test.go
package crypto

// KAT против schema/test-vectors/vectors.json — те же векторы, что проходит
// Kotlin-реализация (messenger-crypto). Паритет двух реализаций и есть контракт.

import (
	"bytes"
	"crypto/ed25519"
	"crypto/mlkem"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"io"
	"os"
	"testing"

	"golang.org/x/crypto/curve25519"
	"golang.org/x/crypto/hkdf"
	"golang.org/x/crypto/nacl/box"
	"golang.org/x/crypto/nacl/secretbox"
)

func loadVectors(t *testing.T) map[string]json.RawMessage {
	t.Helper()
	raw, err := os.ReadFile("../../../schema/test-vectors/vectors.json")
	if err != nil {
		t.Fatalf("vectors.json не найден: %v", err)
	}
	var file struct {
		Vectors map[string]json.RawMessage `json:"vectors"`
	}
	if err := json.Unmarshal(raw, &file); err != nil {
		t.Fatalf("vectors.json не парсится: %v", err)
	}
	return file.Vectors
}

func vec[T any](t *testing.T, vectors map[string]json.RawMessage, name string) T {
	t.Helper()
	rawVec, ok := vectors[name]
	if !ok {
		t.Fatalf("вектор %q отсутствует", name)
	}
	var v T
	if err := json.Unmarshal(rawVec, &v); err != nil {
		t.Fatalf("вектор %q не парсится: %v", name, err)
	}
	return v
}

func unhex(t *testing.T, s string) []byte {
	t.Helper()
	b, err := hex.DecodeString(s)
	if err != nil {
		t.Fatalf("плохой hex: %v", err)
	}
	return b
}

func assertHex(t *testing.T, want string, got []byte, what string) {
	t.Helper()
	if hex.EncodeToString(got) != want {
		t.Fatalf("%s разошёлся с эталоном:\n want %s\n got  %s", what, want, hex.EncodeToString(got))
	}
}

func TestCanonicalBytesVector(t *testing.T) {
	vectors := loadVectors(t)
	v := vec[struct {
		Inputs struct {
			FormatVersion   uint32 `json:"format_version"`
			MessageID       uint64 `json:"message_id"`
			ChatID          string `json:"chat_id"`
			SenderID        string `json:"sender_id"`
			SenderDevice    string `json:"sender_device"`
			Kind            uint32 `json:"kind"`
			CreatedAtUnixMs int64  `json:"created_at_unix_ms"`
			ReplyTo         uint64 `json:"reply_to"`
		} `json:"inputs"`
		CanonicalBytesHex string `json:"canonical_bytes_hex"`
		Sha256Hex         string `json:"sha256_hex"`
	}](t, vectors, "canonical_bytes")

	sb := vec[struct {
		KodiumOutputHex string `json:"kodium_output_hex"`
	}](t, vectors, "secretbox")
	bw := vec[struct {
		EphPublic string `json:"eph_public"`
	}](t, vectors, "box_wrap")

	// Блобы — как в генераторе: payload из secretbox, escrow = 0xa1×1088 ‖ 0xa2×48
	escrowBytes := append(bytes.Repeat([]byte{0xa1}, 1088), bytes.Repeat([]byte{0xa2}, 48)...)
	cb := CanonicalBytes(v.Inputs.FormatVersion, EnvelopeMeta{
		MessageID:       v.Inputs.MessageID,
		ChatID:          v.Inputs.ChatID,
		SenderID:        v.Inputs.SenderID,
		SenderDevice:    v.Inputs.SenderDevice,
		Kind:            v.Inputs.Kind,
		CreatedAtUnixMs: v.Inputs.CreatedAtUnixMs,
		ReplyTo:         v.Inputs.ReplyTo,
	}, unhex(t, sb.KodiumOutputHex), escrowBytes, unhex(t, bw.EphPublic), nil)

	assertHex(t, v.CanonicalBytesHex, cb, "canonical_bytes")
	digest := sha256.Sum256(cb)
	assertHex(t, v.Sha256Hex, digest[:], "sha256(canonical_bytes)")
}

func TestEd25519Vector(t *testing.T) {
	v := vec[struct {
		Seed         string `json:"seed"`
		PublicKey    string `json:"public_key"`
		MessageHex   string `json:"message_hex"`
		SignatureHex string `json:"signature_hex"`
	}](t, loadVectors(t), "ed25519")

	key := ed25519.NewKeyFromSeed(unhex(t, v.Seed))
	assertHex(t, v.PublicKey, key.Public().(ed25519.PublicKey), "ed25519 public key")

	sig := ed25519.Sign(key, unhex(t, v.MessageHex))
	assertHex(t, v.SignatureHex, sig, "ed25519 signature")

	if !VerifyEnvelopeSignature(unhex(t, v.PublicKey), unhex(t, v.MessageHex), sig) {
		t.Fatal("VerifyEnvelopeSignature обязан принять валидную подпись")
	}
	tampered := append([]byte{}, unhex(t, v.MessageHex)...)
	tampered[0] ^= 1
	if VerifyEnvelopeSignature(unhex(t, v.PublicKey), tampered, sig) {
		t.Fatal("VerifyEnvelopeSignature принял подпись повреждённого сообщения")
	}
}

func TestSecretboxVector(t *testing.T) {
	v := vec[struct {
		Key             string `json:"key"`
		Nonce           string `json:"nonce"`
		PlaintextHex    string `json:"plaintext_hex"`
		KodiumOutputHex string `json:"kodium_output_hex"`
	}](t, loadVectors(t), "secretbox")

	var key [32]byte
	var nonce [24]byte
	copy(key[:], unhex(t, v.Key))
	copy(nonce[:], unhex(t, v.Nonce))
	out := secretbox.Seal(nonce[:], unhex(t, v.PlaintextHex), &nonce, &key) // nonce‖box, как у Kodium
	assertHex(t, v.KodiumOutputHex, out, "secretbox nonce‖box")
}

func TestBoxWrapVector(t *testing.T) {
	v := vec[struct {
		EphSecret       string `json:"eph_secret"`
		EphPublic       string `json:"eph_public"`
		RecipientPublic string `json:"recipient_public"`
		Nonce           string `json:"nonce"`
		MessageKey      string `json:"message_key"`
		KodiumOutputHex string `json:"kodium_output_hex"`
	}](t, loadVectors(t), "box_wrap")

	ephPub, err := curve25519.X25519(unhex(t, v.EphSecret), curve25519.Basepoint)
	if err != nil {
		t.Fatal(err)
	}
	assertHex(t, v.EphPublic, ephPub, "X25519 public из секрета")

	var secret, recipientPub [32]byte
	var nonce [24]byte
	copy(secret[:], unhex(t, v.EphSecret))
	copy(recipientPub[:], unhex(t, v.RecipientPublic))
	copy(nonce[:], unhex(t, v.Nonce))
	out := box.Seal(nonce[:], unhex(t, v.MessageKey), &nonce, &recipientPub, &secret)
	assertHex(t, v.KodiumOutputHex, out, "box_wrap nonce‖box")
}

func TestHKDFVector(t *testing.T) {
	v := vec[struct {
		IKM       string `json:"ikm"`
		Salt      string `json:"salt"`
		InfoHex   string `json:"info_hex"`
		OutputHex string `json:"output_hex"`
	}](t, loadVectors(t), "hkdf_sha256")

	out := make([]byte, 32)
	r := hkdf.New(sha256.New, unhex(t, v.IKM), unhex(t, v.Salt), unhex(t, v.InfoHex))
	if _, err := io.ReadFull(r, out); err != nil {
		t.Fatal(err)
	}
	assertHex(t, v.OutputHex, out, "hkdf_sha256")
}

func TestMediaChunkKeysVector(t *testing.T) {
	v := vec[struct {
		MediaKey string `json:"media_key"`
		Chunk0   string `json:"chunk_0"`
		Chunk1   string `json:"chunk_1"`
		Chunk10  string `json:"chunk_10"`
	}](t, loadVectors(t), "media_chunk_keys")

	derive := func(info string) []byte {
		out := make([]byte, 32)
		r := hkdf.New(sha256.New, unhex(t, v.MediaKey), nil, []byte(info)) // salt пустой → нули
		if _, err := io.ReadFull(r, out); err != nil {
			t.Fatal(err)
		}
		return out
	}
	assertHex(t, v.Chunk0, derive("chunk:0"), "chunk_key[0]")
	assertHex(t, v.Chunk1, derive("chunk:1"), "chunk_key[1]")
	assertHex(t, v.Chunk10, derive("chunk:10"), "chunk_key[10]")
}

func TestMLKEM768Vector(t *testing.T) {
	v := vec[struct {
		KeygenSeed      string `json:"keygen_seed"`
		PublicKeyLen    int    `json:"public_key_len"`
		CiphertextLen   int    `json:"ciphertext_len"`
		SharedLen       int    `json:"shared_len"`
		PublicKeySha256 string `json:"public_key_sha256"`
	}](t, loadVectors(t), "mlkem768_escrow")

	dk, err := mlkem.NewDecapsulationKey768(unhex(t, v.KeygenSeed)) // seed = d(32) ‖ z(32)
	if err != nil {
		t.Fatal(err)
	}
	ekBytes := dk.EncapsulationKey().Bytes()
	if len(ekBytes) != v.PublicKeyLen {
		t.Fatalf("длина ek: ожидалось %d, получено %d", v.PublicKeyLen, len(ekBytes))
	}
	digest := sha256.Sum256(ekBytes)
	assertHex(t, v.PublicKeySha256, digest[:], "sha256(ml-kem ek) — keygen(seed)")

	// Инвариант: encapsulate → decapsulate восстанавливает shared (ct рандомизирован)
	shared, ct := dk.EncapsulationKey().Encapsulate()
	if len(ct) != v.CiphertextLen || len(shared) != v.SharedLen {
		t.Fatalf("размеры ct/shared: %d/%d", len(ct), len(shared))
	}
	shared2, err := dk.Decapsulate(ct)
	if err != nil || !bytes.Equal(shared, shared2) {
		t.Fatalf("decapsulate не восстановил shared (err=%v)", err)
	}
}

File: messenger-crypto/src/test/kotlin/io/tima/crypto/VectorsTest.kt
package io.tima.crypto

import io.kodium.KodiumPrivateKey
import io.kodium.core.nacl
import io.kodium.ratchet.HKDF
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.int
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.long
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue
import kotlin.test.fail

/**
 * KAT против канонических векторов schema/test-vectors/vectors.json.
 * Эталон — tweetnacl-js + noble; Kodium обязан совпасть байт-в-байт.
 * Расхождение = красный билд, а не «подстроить тест».
 */
class VectorsTest {

    private val vectors = Json.parseToJsonElement(
        javaClass.getResourceAsStream("/vectors.json")?.readBytes()?.decodeToString()
            ?: fail("vectors.json не найден в тест-ресурсах (../schema/test-vectors)"),
    ).jsonObject["vectors"]!!.jsonObject

    private fun vector(name: String) = vectors[name]?.jsonObject ?: fail("Вектор '$name' отсутствует")
    private fun kotlinx.serialization.json.JsonObject.hex(field: String) =
        this[field]!!.jsonPrimitive.content.hexToBytes()

    @Test
    fun `secretbox - конверт с фиксированным nonce совпадает байт-в-байт`() {
        val v = vector("secretbox")
        val out = EnvelopeCipher.sealWithNonce(v.hex("key"), v.hex("plaintext_hex"), v.hex("nonce"))
        assertEquals(v["kodium_output_hex"]!!.jsonPrimitive.content, out.toHex())

        // Обратный ход боевым API
        val opened = EnvelopeCipher.open(v.hex("key"), out).getOrThrow()
        assertEquals(v["plaintext_hex"]!!.jsonPrimitive.content, opened.toHex())
    }

    @Test
    fun `box_wrap - wrapped_key с фиксированным nonce совпадает байт-в-байт`() {
        val v = vector("box_wrap")

        // Публичный ключ выводится из секрета так же, как в эталоне
        val (ephPub, _) = nacl.Box.keyPairFromSecretKey(v.hex("eph_secret"))
        assertEquals(v["eph_public"]!!.jsonPrimitive.content, ephPub.toHex())

        val out = WrappedKeyService.wrapWithNonce(
            ephemeralSecret = v.hex("eph_secret"),
            recipientPub = v.hex("recipient_public"),
            messageKey = v.hex("message_key"),
            nonce = v.hex("nonce"),
        )
        assertEquals(v["kodium_output_hex"]!!.jsonPrimitive.content, out.toHex())

        // Разворачивание боевым API со стороны получателя
        val recipientKey = KodiumPrivateKey.fromRaw(v.hex("recipient_secret"))
        val unwrapped = WrappedKeyService.unwrap(recipientKey, v.hex("eph_public"), out).getOrThrow()
        assertEquals(v["message_key"]!!.jsonPrimitive.content, unwrapped.toHex())
    }

    @Test
    fun `ed25519 - подпись из seed устройства совпадает байт-в-байт`() {
        val v = vector("ed25519")
        val deviceKey = KodiumPrivateKey.fromRaw(v.hex("seed"))
        assertEquals(v["public_key"]!!.jsonPrimitive.content, deviceKey.getPublicKey().signingKey.toHex())

        val message = v.hex("message_hex")
        val signature = MessageSigner.sign(deviceKey, message).getOrThrow()
        assertEquals(v["signature_hex"]!!.jsonPrimitive.content, signature.toHex())

        assertTrue(MessageSigner.verify(v.hex("public_key"), message, signature))
        // Повреждённое сообщение — подпись невалидна
        val tampered = message.copyOf().also { it[0] = (it[0].toInt() xor 1).toByte() }
        assertTrue(!MessageSigner.verify(v.hex("public_key"), tampered, signature))
    }

    @Test
    fun `hkdf_sha256 - деривация escrow-ключа совпадает байт-в-байт`() {
        val v = vector("hkdf_sha256")
        val out = HKDF.deriveSecrets(salt = v.hex("salt"), ikm = v.hex("ikm"), info = v.hex("info_hex"), length = 32)
        assertEquals(v["output_hex"]!!.jsonPrimitive.content, out.toHex())
    }

    @Test
    fun `canonical_bytes - preimage подписи совпадает байт-в-байт`() {
        val v = vector("canonical_bytes")
        val inputs = v["inputs"]!!.jsonObject

        val meta = EnvelopeMeta(
            messageId = inputs["message_id"]!!.jsonPrimitive.long.toULong(),
            chatId = inputs["chat_id"]!!.jsonPrimitive.content,
            senderId = inputs["sender_id"]!!.jsonPrimitive.content,
            senderDevice = inputs["sender_device"]!!.jsonPrimitive.content,
            kind = inputs["kind"]!!.jsonPrimitive.int,
            createdAtUnixMs = inputs["created_at_unix_ms"]!!.jsonPrimitive.long,
            replyTo = inputs["reply_to"]!!.jsonPrimitive.long.toULong(),
        )
        // Блобы — как в генераторе: payload из secretbox-вектора, escrow — 0xa1/0xa2, eph — из box_wrap
        val encryptedPayload = vector("secretbox").hex("kodium_output_hex")
        val escrowBytes = ByteArray(1088) { 0xa1.toByte() } + ByteArray(48) { 0xa2.toByte() }
        val senderEphemeralPub = vector("box_wrap").hex("eph_public")

        val cb = CanonicalBytes.build(
            meta = meta,
            encryptedPayload = encryptedPayload,
            escrowBytes = escrowBytes,
            senderEphemeralPub = senderEphemeralPub,
            ratchetEnvelope = CanonicalBytes.EMPTY,
            formatVersion = inputs["format_version"]!!.jsonPrimitive.int,
        )
        assertEquals(v["canonical_bytes_hex"]!!.jsonPrimitive.content, cb.toHex())
        assertEquals(v["sha256_hex"]!!.jsonPrimitive.content, sha256(cb).toHex())
    }

    @Test
    fun `mlkem768 - keygen из seed детерминирован, escrow round-trip`() {
        val v = vector("mlkem768_escrow")
        val seed = v.hex("keygen_seed") // 64 байта, layout noble/FIPS 203: d(32) ‖ z(32)

        // Провайдер escrow — Mlkem768 (BouncyCastle). Kodium ML-KEM не интероперабелен
        // с FIPS 203 и этот вектор НЕ проходит — см. Mlkem768.kt и поправку к ADR-0005.
        val (publicKey, secretKey) = Mlkem768.keyPairFromSeed(seed)

        assertEquals(v["public_key_len"]!!.jsonPrimitive.int, publicKey.size)
        assertEquals(v["secret_key_len"]!!.jsonPrimitive.int, secretKey.size)
        assertEquals(
            v["public_key_sha256"]!!.jsonPrimitive.content,
            sha256(publicKey).toHex(),
            "keygen(seed) разошёлся с noble — реализация ML-KEM некорректна",
        )

        // Инвариант (ct рандомизирован по FIPS 203): decapsulate(ct, sk) == shared
        val (shared, ct) = Mlkem768.encapsulate(publicKey)
        assertEquals(v["ciphertext_len"]!!.jsonPrimitive.int, ct.size)
        assertEquals(v["shared_len"]!!.jsonPrimitive.int, shared.size)
        val shared2 = Mlkem768.decapsulate(ct, secretKey) ?: fail("decapsulate вернул null")
        assertEquals(shared.toHex(), shared2.toHex())
    }

    @Test
    fun `все восемь векторов присутствуют`() {
        val expected = setOf(
            "secretbox", "box_wrap", "ed25519", "hkdf_sha256", "media_chunk_keys",
            "canonical_bytes", "message_body", "mlkem768_escrow",
        )
        assertEquals(expected, vectors.keys, "Состав vectors.json изменился — обнови KAT")
    }
}
